ibc/src/component/connection/stateful.rs

```diff
@@ -0,0 +1,496 @@
+    use super::super::*;
+
+    #[async_trait]
+    pub trait ConnectionOpenInitCheck: StateExt {
+        async fn validate(&self, msg: &MsgConnectionOpenInit) -> anyhow::Result<()> {
+            // check that the client with the specified ID exists
+            self.get_client_data(&msg.client_id).await?;
+
+            Ok(())
+        }
+    }
+
+    impl<T: StateExt> ConnectionOpenInitCheck for T {}
+}
+
+pub mod connection_open_confirm {
+    use super::super::*;
+
+    #[async_trait]
+    pub trait ConnectionOpenConfirmCheck: StateExt + inner::Inner {
+        // Validate a ConnectionOpenConfirm message, completing the IBC connection handshake.
+        //
+        // Verify that we have a connection in the correct state (TRYOPEN), and that the
+        // counterparty has committed a connection with the expected state (OPEN) to their state
+        // store.
+        //
+        // Here we are Chain B.
+        // CHAINS:          (A, B)
+        // PRIOR STATE:     (OPEN, TRYOPEN)
+        // POSTERIOR STATE: (OPEN, OPEN)
+        async fn validate(&self, msg: &MsgConnectionOpenConfirm) -> anyhow::Result<()> {
+            // verify that a connection with the provided ID exists and is in the correct state
+            // (TRYOPEN)
+            let connection = self.verify_previous_connection(msg).await?;
+
+            let expected_conn = ConnectionEnd::new(
+                ConnectionState::Open,
+                connection.counterparty().client_id().clone(),
+                Counterparty::new(
+                    connection.client_id().clone(),
+                    Some(msg.connection_id.clone()),
+                    COMMITMENT_PREFIX.as_bytes().to_vec().try_into().unwrap(),
+                ),
+                connection.versions().to_vec(),
+                connection.delay_period(),
+            );
+
+            // get the trusted client state for the counterparty
+            let trusted_client_state = self
+                .get_client_data(connection.client_id())
+                .await?
+                .client_state
+                .0;
+
+            // check if the client is frozen
+            // TODO: should we also check if the client is expired here?
+            if trusted_client_state.is_frozen() {
+                return Err(anyhow::anyhow!("client is frozen"));
+            }
+
+            // get the stored consensus state for the counterparty
+            let trusted_consensus_state = self
+                .get_verified_consensus_state(msg.proofs.height(), connection.client_id().clone())
+                .await?
+                .0;
+
+            let client_def = AnyClient::from_client_type(trusted_client_state.client_type());
+
+            // PROOF VERIFICATION
+            // note: here we only verify connection state inclusion, not client state or consensus
+            // state.
+            //
+            // 1. verify that the counterparty chain committed the expected_conn to its state
```

Comment:
Are those verified somewhere else? Could the comment point to where, if so?

If there's only one step, is it better not to number it, so as not to give the impression there are also steps 2,3, .... ?

Reply:
yep, there's only one step here. i'll write this comment so it's a bit clearer